util/dvsim/Deploy.py

```diff
@@ -279,118 +277,138 @@ def dispatch_cmd(self):
-                    max_odirs = self.max_odirs
-                num_rm_dirs = num_dirs - max_odirs
-                if num_rm_dirs > -1:
-                    rm_dirs = run_cmd(find_cmd +
-                                      "-printf '%T+ %p\n' | sort | head -n " +
-                                      str(num_rm_dirs + 1) +
-                                      " | awk '{print $2}'")
-                    rm_dirs = rm_dirs.replace('\n', ' ')
-                    dirs = dirs.replace(rm_dirs, "")
-                    os.system("/bin/rm -rf " + rm_dirs)
-        except IOError:
-            log.error("Failed to delete old run directories!")
-        return dirs
+        if not os.path.exists(odir):
+            return []
+
+        # If output directory exists, back it up.
+        ts = datetime.fromtimestamp(os.stat(odir).st_ctime)
+        ts = ts.strftime(self.sim_cfg.ts_format)
+        shutil.move(odir, odir + "_" + ts)
+
+        # Get list of past output directories sorted by creation time.
+        pdir = Path(odir).resolve().parent
+        dirs = sorted([old for old in pdir.iterdir() if old.is_dir()],
+                      key=os.path.getctime,
+                      reverse=True)
+
+        for old in dirs[self.max_odirs - 1:]:
+            if os.path.exists(old):
+                shutil.rmtree(old)
+
+        return dirs[0:self.max_odirs - 2]
 
     def _test_passed(self):
-        '''Return True if the job passed, False otherwise
+        """Determine the outcome of the job (P/F if it ran to completion).
 
-        This is called by poll() just after the job finishes.
+        Return True if the job passed, False otherwise. This is called by
+        poll() just after the job finishes.
+
+        """
 
-        '''
         if self.dry_run:
             return True
 
-        seen_fail_pattern = False
-        for fail_pattern in self.fail_patterns:
-            # Return error message with the following 4 lines.
-            grep_cmd = "grep -m 1 -A 4 -E \'" + fail_pattern + "\' " + self.log
-            (status, rslt) = subprocess.getstatusoutput(grep_cmd)
-            if rslt:
-                msg = "```\n{}\n```\n".format(rslt)
-                self.fail_msg += msg
-                log.log(VERBOSE, msg)
-                seen_fail_pattern = True
-                break
-
-        if seen_fail_pattern:
+        # Only one pattern need to be seen.
+        failed = False
+        chk_failed = bool(self.fail_patterns)
+
+        # All pass patterns need to be seen - hence maintain it as bool dict.
+        passed = {p: False for p in self.pass_patterns}
+        chk_passed = bool(self.pass_patterns)
+
+        try:
+            f = open(self.log, "r", encoding="UTF-8")
+        except Exception as e:
+            msg = "Error opening file \"{!r}\":\n{!r}".format(self.log, e)
+            self.fail_msg += msg
+            log.log(VERBOSE, msg)
+            return False
+
+        lines = f.readlines()
```

Comment:
This leaks an open file, I think. Also, catching `Exception` is always a bit dubious and I think your error message won't be formatted right (extra quotes on the filename and repr on an Exception object). Maybe something like this would be better?
```python3
        try:
            with open(self.log, "r", encoding="UTF-8") as f:
                lines = f.readlines()
        except OSError as e:
            msg = "Error opening file {!r}:\n{}".format(self.log, e)
            self.fail_msg += msg
            log.log(VERBOSE, msg)
            return False
```